File: src/doc_analyse/classifiers/openai.py
from __future__ import annotations

from collections.abc import Sequence
from typing import Any, Dict, Optional

from doc_analyse.classifiers.base import (
    DEFAULT_SYSTEM_PROMPT,
    BaseClassifier,
    ClassifierDependencyError,
    ClassifierMessage,
    ensure_api_key,
    require_text_response,
)


class OpenAIClassifier(BaseClassifier):
    provider_name = "openai"
    default_model = "gpt-4o-mini"

    def __init__(
        self,
        model: Optional[str] = None,
        api_key: Optional[str] = None,
        client: Optional[Any] = None,
        temperature: float = 0.0,
        max_tokens: int = 1200,
        system_prompt: str = DEFAULT_SYSTEM_PROMPT,
        request_options: Optional[Dict[str, Any]] = None,
        **client_options: Any,
    ) -> None:
        super().__init__(
            model=model,
            temperature=temperature,
            max_tokens=max_tokens,
            system_prompt=system_prompt,
        )
        self.api_key = api_key
        self._client = client
        self.client_options = client_options
        self.request_options = request_options or {}

    def _complete(self, messages: Sequence[ClassifierMessage]) -> str:
        client = self._get_client()
        payload = [{"role": message.role, "content": message.content} for message in messages]

        if hasattr(client, "responses"):
            response = client.responses.create(
                model=self.model,
                input=payload,
                temperature=self.temperature,
                max_output_tokens=self.max_tokens,
                **self.request_options,
            )
            return _openai_response_text(response)

        response = client.chat.completions.create(
            model=self.model,
            messages=payload,
            temperature=self.temperature,
            max_tokens=self.max_tokens,
            **self.request_options,
        )
        return _openai_chat_response_text(response)

    def _get_client(self) -> Any:
        if self._client is not None:
            return self._client

        options = dict(self.client_options)
        ensure_api_key("OpenAI", ("OPENAI_API_KEY",), self.api_key, options)

        try:
            from openai import OpenAI
        except ImportError as exc:
            raise ClassifierDependencyError(
                "OpenAIClassifier requires the 'openai' package. Install with: pip install openai"
            ) from exc

        self._client = OpenAI(**options)
        return self._client


def _openai_response_text(response: Any) -> str:
    output_text = getattr(response, "output_text", None)
    if isinstance(output_text, str) and output_text.strip():
        return output_text.strip()

    # The Responses API nests text in output items; normalize all text parts together.
    text_parts = []
    for item in getattr(response, "output", []) or []:
        for content in getattr(item, "content", []) or []:
            text = getattr(content, "text", None)
            if isinstance(text, str) and text.strip():
                text_parts.append(text)

    return require_text_response("OpenAI", "".join(text_parts))


def _openai_chat_response_text(response: Any) -> str:
    try:
        text = response.choices[0].message.content
    except (AttributeError, IndexError, TypeError):
        text = None

    return require_text_response("OpenAI", text)

File: src/classifiers/openai.ts
/**
 * OpenAI LLM classifier.
 * Ported from doc_analyse/classifiers/openai.py
 */

import { BaseClassifier, ClassifierMessage } from "./base.js";

export class OpenAIClassifier extends BaseClassifier {
  readonly provider_name = "openai";
  readonly default_model = "gpt-4o-mini";

  private api_key: string | null;
  private _client: unknown | null = null;
  private readonly client_options: Record<string, unknown>;
  private readonly request_options: Record<string, unknown>;

  constructor({
    model,
    api_key,
    client,
    temperature,
    max_tokens,
    system_prompt,
    user_prompt_template,
    request_options,
    ...client_options
  }: {
    model?: string | null;
    api_key?: string | null;
    client?: unknown;
    temperature?: number | null;
    max_tokens?: number | null;
    system_prompt?: string | null;
    user_prompt_template?: string | null;
    request_options?: Record<string, unknown>;
    [key: string]: unknown;
  } = {}) {
    super({ model, temperature, max_tokens, system_prompt, user_prompt_template });
    this.api_key = api_key ?? null;
    this._client = client ?? null;
    this.client_options = client_options;
    this.request_options = request_options ?? {};
  }

  _complete(messages: readonly ClassifierMessage[]): string {
    const client = this._get_client();
    const payload = messages.map((m) => ({
      role: m.role,
      content: m.content,
    }));

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const c = client as any;
    if (typeof c.responses !== "undefined") {
      const response = c.responses.create({
        model: this.model,
        input: payload,
        temperature: this.temperature,
        max_output_tokens: this.max_tokens,
        ...this.request_options,
      });
      return _openai_response_text(response);
    }

    const response = c.chat.completions.create({
      model: this.model,
      messages: payload,
      temperature: this.temperature,
      max_tokens: this.max_tokens,
      ...this.request_options,
    });
    return _openai_chat_response_text(response);
  }

  private _get_client(): unknown {
    if (this._client !== null) return this._client;

    const options: Record<string, unknown> = { ...this.client_options };
    const { ensure_api_key } = require("./base.js");
    ensure_api_key("OpenAI", ["OPENAI_API_KEY"], this.api_key, options);

    let OpenAI: unknown;
    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      OpenAI = require("openai").OpenAI;
    } catch {
      throw new (require("./base.js").ClassifierDependencyError)(
        "OpenAIClassifier requires the 'openai' package. Install with: npm install openai",
      );
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    this._client = new (OpenAI as any)(options);
    return this._client;
  }
}

function _openai_response_text(response: {
  output_text?: string;
  output?: readonly { content?: readonly { text?: string }[] }[];
}): string {
  const { require_text_response } = require("./base.js");

  if (response.output_text && response.output_text.trim()) {
    return response.output_text.trim();
  }

  const textParts: string[] = [];
  for (const item of response.output ?? []) {
    for (const content of item.content ?? []) {
      if (content.text && content.text.trim()) {
        textParts.push(content.text);
      }
    }
  }

  return require_text_response("OpenAI", textParts.join(""));
}

function _openai_chat_response_text(response: {
  choices?: readonly { message?: { content?: string } }[];
}): string {
  const { require_text_response } = require("./base.js");
  let text: string | null = null;

  try {
    text = response.choices?.[0]?.message?.content ?? null;
  } catch {
    text = null;
  }

  return require_text_response("OpenAI", text);
}
